File: roadmap/java/0001-two-sum.java
package roadmap.java;

import java.util.HashMap;
import java.util.Map;

class Solution {
     public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target-nums[i];
            if (map.containsKey(complement)) {
                int[] result = { map.get(complement), i };
                return result;
                // return new int[]{map.get(complement), i};
            }
            map.put(nums[i], i);
        }

        return null;
    }
}

File: roadmap/go/0001-two-sum.go
package roadmap

// TwoSum return indices of the two numbers such that they add up to a specific target.
// Approach: One-pass Hash Table
// https://leetcode-cn.com/problems/two-sum/
func TwoSum(nums []int, target int) []int {
	preMap := make(map[int]int)
	for i, num := range nums {
		complement := target - num
		if j, ok := preMap[complement]; ok {
			return []int{j, i}
		}
		preMap[num] = i
	}
	return nil
}
